crates/rpc/rpc/src/eth/bundle.rs

```diff
@@ -0,0 +1,205 @@
+//! `Eth` bundle implementation and helpers.
+
+use crate::{
+    eth::{
+        error::{EthApiError, EthResult, RpcInvalidTransactionError},
+        revm_utils::FillableTransaction,
+        utils::recover_raw_transaction,
+        EthTransactions,
+    },
+    BlockingTaskGuard,
+};
+use reth_primitives::{keccak256, U256};
+use reth_revm::database::StateProviderDatabase;
+use reth_rpc_types::{EthCallBundle, EthCallBundleResponse, EthCallBundleTransactionResult};
+use revm::{
+    db::CacheDB,
+    primitives::{Env, ResultAndState, TxEnv},
+};
+use revm_primitives::db::{DatabaseCommit, DatabaseRef};
+use std::sync::Arc;
+
+/// `Eth` bundle implementation.
+pub struct EthBundle<Eth> {
+    /// All nested fields bundled together.
+    inner: Arc<EthBundleInner<Eth>>,
+}
+
+impl<Eth> EthBundle<Eth> {
+    /// Create a new `EthBundle` instance.
+    pub fn new(eth_api: Eth, blocking_task_guard: BlockingTaskGuard) -> Self {
+        Self { inner: Arc::new(EthBundleInner { eth_api, blocking_task_guard }) }
+    }
+}
+
+impl<Eth> EthBundle<Eth>
+where
+    Eth: EthTransactions + 'static,
+{
+    /// Simulates a bundle of transactions at the top of a given block number with the state of
+    /// another (or the same) block. This can be used to simulate future blocks with the current
+    /// state, or it can be used to simulate a past block. The sender is responsible for signing the
+    /// transactions and using the correct nonce and ensuring validity
+    pub async fn call_bundle(&self, bundle: EthCallBundle) -> EthResult<EthCallBundleResponse> {
+        let EthCallBundle { txs, block_number, state_block_number, timestamp } = bundle;
+        if txs.is_empty() {
+            return Err(EthApiError::InvalidParams(
+                EthBundleError::EmptyBundleTransactions.to_string(),
+            ))
+        }
+        if block_number.to::<u64>() == 0 {
+            return Err(EthApiError::InvalidParams(
+                EthBundleError::BundleMissingBlockNumber.to_string(),
+            ))
+        }
+
+        let transactions =
+            txs.into_iter().map(recover_raw_transaction).collect::<Result<Vec<_>, _>>()?;
+
+        let (cfg, mut block_env, at) =
+            self.inner.eth_api.evm_env_at(state_block_number.into()).await?;
+
+        // need to adjust the timestamp for the next block
+        if let Some(timestamp) = timestamp {
+            block_env.timestamp = U256::from(timestamp);
+        } else {
+            block_env.timestamp += U256::from(12);
+        }
+
+        // use the block number of the request
+        block_env.number = U256::from(block_number);
```

Comment:
should we validate that `block_number == state_block_number + 1` or is it not required?

Reply:
there's no such check in the ref impl